src/backend/distributed/commands/table.c

```diff
@@ -593,59 +613,243 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+	List *foreignKeyConstraintList =
+		GetAlterTableAddFKeyConstraintList(alterTableStatement);
+	if (list_length(foreignKeyConstraintList) == 0)
+	{
+		/* we are not defining any foreign keys */
+		return false;
+	}
 
-	LOCKMODE lockmode = AlterTableGetLockLevel(commandList);
+	List *rightRelationIdList =
+		GetAlterTableAddFKeyRightRelationIdList(alterTableStatement);
+
+	LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
 	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
+	if (!IsCitusTable(leftRelationId))
+	{
+		return RelationIdListContainsCitusTableType(rightRelationIdList,
+													CITUS_TABLE_WITH_NO_DIST_KEY);
+	}
+	else if (IsCitusTableType(leftRelationId, CITUS_TABLE_WITH_NO_DIST_KEY))
+	{
+		return RelationIdListContainsPostgresTable(rightRelationIdList);
+	}
+
+	return false;
+}
+
 
-	if (IsCitusTable(leftRelationId))
+/*
+ * RelationIdListContainsCitusTableType returns true if given relationIdList
+ * contains a citus table with given type.
+ */
+static bool
+RelationIdListContainsCitusTableType(List *relationIdList, CitusTableType citusTableType)
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
 	{
-		/* left relation is not a postgres local table, */
-		return;
+		if (IsCitusTableType(relationId, citusTableType))
+		{
+			return true;
+		}
 	}
 
-	List *alterTableFKeyConstraints =
-		GetAlterTableStmtFKeyConstraintList(alterTableStatement);
-	Constraint *constraint = NULL;
-	foreach_ptr(constraint, alterTableFKeyConstraints)
+	return false;
+}
+
+
+/*
+ * RelationIdListContainsPostgresTable returns true if given relationIdList
+ * contains a postgres table.
+ */
+static bool
+RelationIdListContainsPostgresTable(List *relationIdList)
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
 	{
-		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
-											   alterTableStatement->missing_ok);
-		if (IsCitusTableType(rightRelationId, CITUS_LOCAL_TABLE))
+		if (OidIsValid(relationId) && !IsCitusTable(relationId))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	RangeVar *leftRelationRangeVar = alterTableStatement->relation;
+	List *relationRangeVarList = lappend(rightRelationRangeVarList, leftRelationRangeVar);
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
```

Comment:
That does not really prevents deadlocks, but I suppose it can't hurt.

Reply:
true, it can't prevent many cases but at least could prevent same ddl command from different sessions